sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,574 @@
+
+                // add method that users can call in startup to register the source-generated file
+                AddRegistrationExtension(indentedTextWriter);
+
+                indentedTextWriter.Indent--;
+                indentedTextWriter.WriteLine("}");
+
+                indentedTextWriter.Flush();
+                sourceText = SourceText.From(stringWriter.ToString(), encoding: Encoding.UTF8);
+            }
+
+            // Add the source code to the compilation
+            context.AddSource($"GeneratedFunctionMetadataProvider.g.cs", sourceText);
+        }
+
+        /// <summary>
+        /// Register a factory that can create our custom syntax receiver
+        /// </summary>
+        /// <param name="context"></param>
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
+        }
+
+        /// <summary>
+        /// This method populates a dictionary with argumentName (keys) and argumentValue (value) pairs given attribMethodSymbol, which
+        /// represents the attribute constructor as as an IMethodSymbol, and attributeData which stores constructor argument values.
+        /// </summary>
+        /// <param name="attribMethodSymbol">The attribute's constructor as an IMethodSymbol.</param>
+        /// <param name="attributeData">Contains constructor arguments for the constructor represented in attribMethodSymbol.</param>
+        /// <param name="dict">A dicitonary to be populated with constructor arguments.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal static void LoadConstructorArguments(IMethodSymbol attribMethodSymbol, AttributeData attributeData, IDictionary<string, object> dict)
+        {
+            if (attribMethodSymbol.Parameters.Length < attributeData.ConstructorArguments.Length)
+            {
+                throw new InvalidOperationException($"The constructor at '{nameof(attribMethodSymbol)}' has less total arguments than '{nameof(attributeData)}'.");
+            }
+
+            // It's fair to assume than constructor arguments appear before named arguments, and
+            // that the constructor names would match the property names
+            for (int i = 0; i < attributeData.ConstructorArguments.Length; i++)
+            {
+                var argumentName = attribMethodSymbol.Parameters[i].Name;
+
+                var arg = attributeData.ConstructorArguments[i];
+                switch (arg.Kind)
+                {
+                    case TypedConstantKind.Error:
+                        break;
+                    case TypedConstantKind.Primitive:
+                        dict[argumentName] = arg.Value;
+                        break;
+                    case TypedConstantKind.Enum:
+                        dict[argumentName] = "Enum.GetName(typeof(" + arg.Type!.Name.ToString() + "), " + arg.Value + ")";
```

Comment:
`$"((AuthorizationLevel){arg.Value}).ToString();"` will also produce same value (string version of enum value).

I benchmarked both approaches and this one seems a tiny bit performant.

https://gist.github.com/kshyju/d4318e0b2c1542cb327c0a4090a80f46